src/api/java/appeng/api/crafting/IPatternDetails.java

```diff
@@ -21,21 +21,39 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
-package appeng.api.networking.crafting;
+package appeng.api.crafting;
 
-import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 
-public interface ICraftingCallback {
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
 
-    /**
-     * The call back is synchronized with the level you passed as well as all other crafting jobs. This can lead to
-     * deadlocks in cases like submitting new crafting jobs from the callback itself. However it is not limited to this
-     * case and can deadlock in other ones.
-     * 
-     * It will also be called from a separate crafting thread. If your callback is not thread safe you have to take care
-     * if it yourself. For example by using a thread safe queue and handling it with the blockentity/world/server tick.
-     *
-     * @param job - final job
-     */
-    void calculationComplete(@Nonnull ICraftingJob job);
+import appeng.api.storage.data.IAEStack;
+
+public interface IPatternDetails {
+    ItemStack copyDefinition();
+
+    // TODO: is this even useful?
+    boolean isCrafting();
+
+    IInput[] getInputs();
+
+    default IAEStack getPrimaryOutput() {
```

Comment:
Separate method for only the secondary outputs might also be useful.

Reply:
Might be, but it would require an allocation every time. Probably easier to use the full array and start iteration at 1 if people need it.